app/controllers/blogs_controller.rb

```diff
@@ -9,17 +9,24 @@ def index
     @blogs = Blog.search(params[:term]).published.default_order
   end
 
-  def show; end
+  def show
+    raise ActiveRecord::RecordNotFound if !user_signed_in? && @blog.secret?
+    raise ActiveRecord::RecordNotFound if user_signed_in? && @blog.secret? && @blog.user != current_user
```

Comment:
なるほどー、まず `set_blog` でブログの集合全体から find して、次にこの行で「見せてもよいもの」かどうかチェックして、該当しなければ `ActiveRecord::RecordNotFound` を手動で raise する、ということですね。やりたいことができていそうな点はまず good です:+1: その上で、要件通りに動くだけではなく、より Rails らしく保守性が高いとされる書き方に寄せたいです。

まず、Rails の基本原則として、コントローラはなるべく薄く保つ方が良く、モデルやその他のメソッドを呼ぶ程度にした方が好ましいというのがあります。
また、この行で if 文に書かれている条件を合わせると、「ユーザーがアクセスできてよいのは、公開記事と自分自身の秘密の記事である」という、ブログ記事に関する業務知識が浮かび上がってきます。これは本来コントローラではなく Blog モデルが持っている方がオブジェクト指向的にも自然だったりします。
なので Blog モデルの方に**「見せて良いものの集合」を返す何か**を実装しておいて、ここからはそれに対して find するだけ、という形が好ましいです〜！

別の観点から言うと、find って引数に渡した id に対応するレコードが見つからないときに ActiveRecord::RecordNotFound を上げてくれるんですよね。

[https://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find](https://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find)

> Find by id - This can either be a specific id (1), a list of ids (1, 5, 6), or an array of ids ([5, 6, 10]). If one or more records cannot be found for the requested ids, then ActiveRecord::RecordNotFound will be raised.

なので、その意味でもなにかに対して find させて、見つからなければ自然と ActiveRecord::RecordNotFound が発生するように、というのが正攻法だったりします！（つまりここは出題者からの「`find` が raise するほうの ActiveRecord::RecordNotFound を使ってくれぇ〜」というメッセージだったわけですね🙂）